### PR TITLE
Fix issues with `type/PK` + `type/DateTime`  fields

### LINKED ISSUE #6290: Raw queries with "ON" datefilter are broken for Druid [Regression]
Hi,

Since metabase 0.26.2, there is an issue with raw queries that have a filter set to ON specific date or time. When filter "on XX/xx/2017" is added only 1 record (i guess first) is shown in the "details"/"1 record" format instead of a table with all records.
This does not happen if filter is changed to past X days for example or current day.

Database: druid

### PR DESCRIPTION
**Pull request by tlrobinson**
- [x] if drilling into a `type/PK` + `type/DateTime` field use `["datetime-field", ..., "default"]` to disable implicit bucketing by day
- [x] `getMode` shouldn't detect `type/PK` + `type/DateTime` fields as PKs (and thus `ObjectMode`) unless `["datetime-field", ..., "default"]` is used
- [x] `getMode` shouldn't return `ObjectMode` if results have more than one row.
- [x] tests

Resolves #6290
Builds on #6724

### CODE REVIEW

**Comment by tlrobinson** [comment]
By the way, the original issue can be reproduced on any database by marking a `DateTime` field as "Entity Key" (i.e. `PK`)

**Comment by attekei** [comment]
I reviewed the source code earlier today and it looks good to me

**Review by camsaul — APPROVED** [review]
LGTM

**Comment by salsakran** [comment]
Ok, so I'm really late to the party -- 
do we need to mark the druid timestamp as a PK at all? I.e. are we just adding a weird bandaid for the type system being incorrect here? 

I don't think a time bucket should be considered a pk personally, but I haven't thought deeply about this.

**Comment by camsaul** [comment]
Yeah I don't think timestamps should be PKs. I don't remember why they were in the first place but it doesn't sound like it makes sense now

**Comment by salsakran** [comment]
imo this entire PR can be replaced by removing that assignment in the druid driver (?) and forcing a re-classification in the next version.

**Comment by tlrobinson** [comment]
Do you think there are ever cases where a date/time field would be considered a PK? If so I think this PR is still useful.

**Comment by tlrobinson** [comment]
@salsakran ^

**Comment by salsakran** [comment]
I don't really know of any case where you'd want a binned time dimension to be a primary key.

**Review by senior — COMMENTED** [review]
@salsakran I've added a note [here](https://github.com/metabase/metabase/issues/6290#issuecomment-361640643) with the fix you've suggested. Should we close this?

**Comment by tlrobinson** [comment]
@salsakran How should a `id TIMESTAMP PRIMARY KEY` (in postgres, for example) field behave? Currently it's detected as both a `DateTime` and `PK` so clicking the "View Details" drilldown bins the `=` filter by day (because it's a `DateTime`) and silently just shows the first row in the results in the object detail (because it's a `PK`).

Also this (ensuring rows.length == 1 for object detail) fixes another edge case where someone accidentally marks a field as `PK` but it's not actually unique and we show the object detail even if multiple rows are returned.

**Comment by tlrobinson** [comment]
@salsakran ping ^

**Comment by camsaul** [comment]
Are we still planning on merging this?

**Comment by tlrobinson** [comment]
I think we should, but still waiting to hear @salsakran's thoughts

**Comment by salsakran** [comment]
I think we should close this and fix the backend marking druid timestamps as PKs.

A timestamp is most certainly not a PK as classically defined and I don't see a need to expand the definition in this case.

I can be convinced otherwise, but that's my ~80% conviction belief.

**Comment by tlrobinson** [comment]
None of the code in this PR should be merged? It addresses multiple issues.

**Review by salsakran — CHANGES_REQUESTED** [review]
@cam what does "default" do on a `datatime-field` clause? Is it the minimum resolution or day?

File: src/metabase/driver/druid/query_processor.clj [hunk 1]
```diff
@@ -586,28 +586,32 @@
...
-      ;; <= "2015-12-11" -> ["-5000/2015-12-12"]
-      :<=       (make-intervals nil (i/add-date-time-units value 1))
-      ;; This is technically allowed by the QL here but doesn't make sense since every Druid event has a timestamp.
-      ;; Just ignore it
-      :is-null  (log/warn (u/format-color 'red (str "WARNING: timestamps can never be nil. Ignoring IS_NULL filter "
-                                                    "for timestamp.")))
-      ;; :timestamp is always non-nil so nothing to do here
-      :not-null nil)))
+    (let [value (update-in value [:field :unit] (fn [unit]
+                                                  (if (= unit :default)
```
[the remark this answers is not in the document]
Comment (salsakran) [reply src/metabase/driver/druid/query_processor.clj]: @camsaul

**Comment by salsakran** [comment]
also @camsaul how can we do an exact filter (i.e. without any binning) on a datatime column?

**Comment by camsaul** [comment]
@salsakran in answer to both questions see https://github.com/metabase/metabase/wiki/Query-Language-'98#datetime-units. `:default` means don't do any bucketing and use whatever resolution the column is at (e.g. millisecond)

**Comment by tlrobinson** [comment]
The MBQL documentation doesn't match what we actually do. It says 

> Datetime fields used without being wrapped in a datetime-field form are automatically converted to datetime-field forms with :default bucketing.

https://github.com/metabase/metabase/wiki/Query-Language-'98#datetime-fields

> default | N/A | 2016-02-01 21:59:20 | No bucketing should take place -- use the default resolution of the column in question, such as seconds or milliseconds.

https://github.com/metabase/metabase/wiki/Query-Language-'98#datetime-units

However, in many (most? all?) of our drivers if you don't wrap a date time field in a `datetime-field` we bucket by day, which contradicts at least on of those statements.

**Comment by camsaul** [comment]
Oh yeah the first statement is wrong it should say `:day` and not `:default`

File: src/metabase/driver/druid/query_processor.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/driver/druid/query_processor.clj]: done

**Comment by camsaul** [comment]
This is way out of date motion to close

**Comment by camsaul** [comment]
Closing this and other old POC or WIP PRs for now to help clear the PR backlog. I'll leave the branch open in case we want to revisit this

**Pull request closed without merging**